File: src/types/import.ts
export type SourceFileType =
  | "pdf"
  | "document"
  | "presentation"
  | "spreadsheet"
  | "markdown"
  | "text"
  | "image"
  | "html"
  | "csv"
  | "url"
  | "unknown";

export type ExtractionStatus =
  | "pending"
  | "extracted"
  | "unsupported"
  | "failed";

export type ConflictType =
  | "exact_duplicate"
  | "name_collision"
  | "path_conflict"
  | "unknown";

export type ConflictResolution =
  | "skip"
  | "link_to_existing"
  | "rename"
  | "overwrite";

export interface SourceMetadata {
  title: string | null;
  author: string | null;
  created: string | null;
  modified: string | null;
  pageCount: number | null;
  wordCount: number | null;
  language: string | null;
}

export interface ImportConflict {
  originalName: string;
  conflictType: ConflictType;
  existingPath: string | null;
  resolvedPath: string;
  existingHash: string | null;
  newHash: string;
  resolution: ConflictResolution | null;
}

export interface ImportFileEntry {
  originalName: string;
  sourcePath: string;
  archivedPath: string;
  fileType: SourceFileType;
  sizeBytes: number;
  hash: string;
  extractionStatus: ExtractionStatus;
  extractionError: string | null;
  textPreview: string | null;
  pageCount: number | null;
  wordCount: number | null;
  metadata: SourceMetadata | null;
  extractedTextPath: string | null;
  extractedAssets: string[];
  conflict: ImportConflict | null;
  renamedFrom: string | null;
}

export interface ImportSummary {
  totalFiles: number;
  archivedFiles: number;
  duplicateFiles: number;
  renamedFiles: number;
  failedFiles: number;
  conflictsCount: number;
}

export interface ImportPreview {
  files: ImportFileEntry[];
  conflicts: ImportConflict[];
  summary: ImportSummary;
}

export interface ExtractResult {
  originalName: string;
  fileType: SourceFileType;
  status: ExtractionStatus;
  error: string | null;
  textPreview: string | null;
  metadata: SourceMetadata | null;
  extractedTextPath: string | null;
  extractedAssets: string[];
}

export interface ConfirmedImport {
  preview: ImportPreview;
  confirmedAt: string;
}

export interface ImportPreviewRequest {
  projectId: string;
  projectRootPath: string;
  sourcePaths: string[];
  allowDuplicates: boolean;
  linkDuplicates: boolean;
}

export interface ConfirmImportRequest {
  projectId: string;
  projectRootPath: string;
  preview: ImportPreview;
  /** When true, the backend creates a scoped Git checkpoint of the archived
   * files plus source/conflict indices. Backward-compatible default: false. */
  createCheckpoint?: boolean;
}

export interface ExtractTextRequest {
  projectId: string;
  projectRootPath: string;
  sourcePath: string;
}

export interface ValidateUrlRequest {
  url: string;
}

export interface PreviewTextImportRequest {
  projectId: string;
  projectRootPath: string;
  kind: "clipboard" | "url";
  sourceName: string;
  content: string;
  title: string | null;
  author: string | null;
}

export interface FetchedImportUrl {
  url: string;
  html: string;
}

export interface ImportedSource {
  path: string;
  sizeBytes: number;
  fileType: SourceFileType;
}

export const FILE_TYPE_LABELS: Record<SourceFileType, string> = {
  pdf: "PDF",
  document: "Document",
  presentation: "Presentation",
  spreadsheet: "Spreadsheet",
  markdown: "Markdown",
  text: "Text",
  image: "Image",
  html: "HTML",
  csv: "CSV",
  url: "URL",
  unknown: "Unknown",
};

export const EXTRACTION_STATUS_LABELS: Record<ExtractionStatus, string> = {
  pending: "Pending",
  extracted: "Extracted",
  unsupported: "Unsupported",
  failed: "Failed",
};

File: src-tauri/src/models/import.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportRequest {
    pub source_paths: Vec<String>,
    pub allow_duplicates: bool,
    pub link_duplicates: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportPreview {
    pub files: Vec<ImportFileEntry>,
    pub conflicts: Vec<ImportConflict>,
    pub summary: ImportSummary,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub v2_session_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportFileEntry {
    pub original_name: String,
    pub source_path: String,
    pub archived_path: String,
    pub file_type: SourceFileType,
    pub size_bytes: u64,
    pub hash: String,
    pub extraction_status: ExtractionStatus,
    pub extraction_error: Option<String>,
    pub text_preview: Option<String>,
    pub page_count: Option<u32>,
    pub word_count: Option<u32>,
    pub metadata: Option<SourceMetadata>,
    #[serde(default)]
    pub extracted_text_path: Option<String>,
    #[serde(default)]
    pub extracted_assets: Vec<String>,
    pub conflict: Option<ImportConflict>,
    pub renamed_from: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum SourceFileType {
    Pdf,
    Document,
    Presentation,
    Spreadsheet,
    Markdown,
    Text,
    Image,
    Html,
    Csv,
    Url,
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ExtractionStatus {
    Pending,
    Extracted,
    Unsupported,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceMetadata {
    pub title: Option<String>,
    pub author: Option<String>,
    pub created: Option<String>,
    pub modified: Option<String>,
    pub page_count: Option<u32>,
    pub word_count: Option<u32>,
    pub language: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportConflict {
    pub original_name: String,
    pub conflict_type: ConflictType,
    pub existing_path: Option<String>,
    pub resolved_path: String,
    pub existing_hash: Option<String>,
    pub new_hash: String,
    pub resolution: Option<ConflictResolution>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConflictType {
    ExactDuplicate,
    NameCollision,
    PathConflict,
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ConflictResolution {
    Skip,
    LinkToExisting,
    Rename,
    Overwrite,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportSummary {
    pub total_files: u32,
    pub archived_files: u32,
    pub duplicate_files: u32,
    pub renamed_files: u32,
    pub failed_files: u32,
    pub conflicts_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ExtractResult {
    pub original_name: String,
    pub file_type: SourceFileType,
    pub status: ExtractionStatus,
    pub error: Option<String>,
    pub text_preview: Option<String>,
    pub metadata: Option<SourceMetadata>,
    pub extracted_text_path: Option<String>,
    pub extracted_assets: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ConfirmedImport {
    pub preview: ImportPreview,
    pub confirmed_at: String,
    /// Commit hash of the scoped Git checkpoint created when the caller
    /// requested `create_checkpoint: true` on confirm. Serialized as JSON
    /// `null` (not omitted) when no checkpoint was created, matching the
    /// `PendingAction.checkpoint_hash` convention so the frontend can use a
    /// uniform `!== null` check.
    #[serde(default)]
    pub checkpoint_hash: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceArtifactIndex {
    #[serde(default)]
    pub sources: std::collections::HashMap<String, Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ImportedSource {
    pub path: String,
    pub size_bytes: u64,
    pub file_type: SourceFileType,
}
